src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs

```diff
@@ -158,34 +158,48 @@ public virtual void SetParameters(ParameterCollection parameters)
                 _hasCalledInit = true;
                 OnInit();
 
-                // If you override OnInitAsync and return a nonnull task, then by default
+                // If you override OnInitAsync and return a noncompleted task, then by default
                 // we automatically re-render once that task completes.
                 var initTask = OnInitAsync();
-                if (initTask != null && initTask.Status != TaskStatus.RanToCompletion)
-                {
-                    initTask.ContinueWith(ContinueAfterLifecycleTask);
-                }
+                ContinueAfterLifecycleTask(initTask);
             }
 
             OnParametersSet();
             var parametersTask = OnParametersSetAsync();
-            if (parametersTask != null && parametersTask.Status != TaskStatus.RanToCompletion)
-            {
-                parametersTask.ContinueWith(ContinueAfterLifecycleTask);
-            }
+            ContinueAfterLifecycleTask(parametersTask);
 
             StateHasChanged();
         }
 
-        private void ContinueAfterLifecycleTask(Task task)
+        private async void ContinueAfterLifecycleTask(Task task)
         {
-            if (task.Exception == null)
+            if (task == null)
             {
-                StateHasChanged();
+                // Treat like a pre-completed task that can't possibly raise errors
+                // i.e., do nothing
+            }
+            else if (task.IsCompleted)
+            {
+                // Since it's already completed synchronously, no need to await and no
+                // need to issue a further render (we already rerender synchronously).
+                // Just need to make sure we propagate any errors.
+                if (task.IsFaulted)
+                {
+                    HandleException(task.Exception);
+                }
             }
             else
             {
-                HandleException(task.Exception);
+                // For incomplete tasks, automatically re-render on completion
+                try
+                {
+                    await task;
```

Comment:
This is a nice imrovment and should fix some issues we were seeing with the execution context.